tests/lib/Controller/Trait/RequestTraitTest.php

```diff
@@ -0,0 +1,184 @@
+    }
+}
+
+// phpcs:ignore PSR1.Classes.ClassDeclaration.MultipleClasses
+class RequestTraitTest extends TestCase
+{
+    use RequestTrait;
+
+    private Request $request;
+    private GenericController $controller;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+        $this->request = new Request();
+        $this->expectedStateAuthId =  OAuth2::AUTHID;
+        $this->controller = $this->getMockBuilder(GenericController::class)
+            ->onlyMethods(['loadState', 'getSourceService'])
+            ->getMock();
+    }
+
+    public function testStateIsValidWithMissingState(): void
+    {
+        $this->assertFalse($this->controller->stateIsValid($this->request));
+    }
+
+    public function testStateIsValidWithEmptyState(): void
+    {
+        $this->request->query->set('state', '');
+        $this->assertFalse($this->controller->stateIsValid($this->request));
+    }
+
+    public function testStateIsValidWithValidState(): void
+    {
+        $this->request->query->set('state', OAuth2::STATE_PREFIX . '|example');
+        $this->assertTrue($this->controller->stateIsValid($this->request));
+    }
+
+    public function testStateIsValidWithInvalidState(): void
+    {
+        $this->request->query->set('state', 'invalid|example');
+        $this->assertFalse($this->controller->stateIsValid($this->request));
+    }
+
+    public function testParseRequestWithInvalidState(): void
+    {
+        $this->expectException(BadRequest::class);
+        $this->expectExceptionMessage('An error occured');
+        $this->request->attributes->set('_route', 'invalid_route');
+        $this->controller->parseRequest($this->request);
+    }
+
+    public function testParseRequestWithEmptyState(): void
+    {
+        $this->request->query->set('state', '');
+        $this->expectException(BadRequest::class);
+        $this->expectExceptionMessage('An error occured');
+        $this->request->attributes->set('_route', 'invalid_route');
+        $this->controller->parseRequest($this->request);
+    }
+
+    public function testParseRequestWithValidState(): void
+    {
+        $this->request->query->set('state', OAuth2::STATE_PREFIX . '|valid_state');
+        $this->request->attributes->set('_route', 'valid_route');
+
+        /** @psalm-suppress UndefinedMethod,MixedMethodCall */
+        $this->controller->method('loadState')->willReturn([OAuth2::AUTHID => 'test_authsource_id']);
+
+        $this->controller->setSourceId('test_authsource_id');
+
+        $mockSourceService = $this->getMockBuilder(SourceService::class)->getMock();
+        $mockSource = $this->getMockBuilder(Source::class)->disableOriginalConstructor()->getMock();
+        /** @psalm-suppress UndefinedMethod,MixedMethodCall */
+        $this->controller->method('getSourceService')->willReturn($mockSourceService);
+
+        $mockSourceService->method('getById')->willReturn($mockSource);
+
+        $this->controller->setSource($mockSource);
```

Comment:
I don't think should be set.  The expectation is the `parseRequest` will set it correct?